src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs

```diff
@@ -277,26 +326,41 @@ public void Load()
 
                 // Compare to UseHttps(httpsOptions => { })
                 var httpsOptions = new HttpsConnectionAdapterOptions();
+
                 if (https)
                 {
-                    httpsOptions.SslProtocols = ConfigurationReader.EndpointDefaults.SslProtocols ?? SslProtocols.None;
-                    httpsOptions.ClientCertificateMode = ConfigurationReader.EndpointDefaults.ClientCertificateMode ?? ClientCertificateMode.NoCertificate;
-
                     // Defaults
                     Options.ApplyHttpsDefaults(httpsOptions);
 
                     if (endpoint.SslProtocols.HasValue)
                     {
                         httpsOptions.SslProtocols = endpoint.SslProtocols.Value;
                     }
+                    else
+                    {
+                        // Ensure endpoint is reloaded if it used the default protocol and the SslProtocols changed.
+                        endpoint.SslProtocols = ConfigurationReader.EndpointDefaults.SslProtocols;
+                    }
 
                     if (endpoint.ClientCertificateMode.HasValue)
                     {
                         httpsOptions.ClientCertificateMode = endpoint.ClientCertificateMode.Value;
                     }
+                    else
+                    {
+                        // Ensure endpoint is reloaded if it used the default mode and the ClientCertificateMode changed.
+                        endpoint.ClientCertificateMode = ConfigurationReader.EndpointDefaults.ClientCertificateMode;
+                    }
 
-                    // Specified
-                    httpsOptions.ServerCertificate = LoadCertificate(endpoint.Certificate, endpoint.Name)
+                    if (endpoint.Sni.Count == 0)
+                    {
+                        // Ensure endpoint is reloaded if it used the default SNI config and it changed.
+                        // No need to configure httpsOptions for SNI since the SniOptionsSelector will now use the EndpointDefaults SNI config.
+                        endpoint.Sni = ConfigurationReader.EndpointDefaults.Sni;
```

Comment:
What's the precedence order here? EndpointDefaults.Sni shouldn't take precedence over Endpoint.Certificate.

Reply:
This is why I was apprehensive about adding EndpointDefaults.Sni to begin with. Endpoint.Certificate doesn't take precedence over Endpoint.Sni. Having precedence where Sni is prioritized over Certificate usually but not always is weird.

If any other property is defined in EndpointDefaults, but not directly on the Endpoint, we always treat it as if the property defined in EndpointDefaults was defined directly on the Endpoint. It doesn't makes sense to me to carve out an exception to this rule just for SNI.

If you want to use the Endpoint.Certificate don't define EndpointDefaults.Sni, or at least don't define it for a server name the endpoint serves and add an empty wildcard ("*") SNI so then it will fall back to the Endpoint.Certificate.

I would rather remove EndpointDefaults.Sni than have some special convoluted fallback logic just for EndpointDefaults.Sni.

Reply:
`EndpointDefaults:Sni` is removed.